---
title: "Basic Bioinformatics Overview"
author: "PennCHOP Microbiome Program"
date: \today
output: 
    pdf_document:
        keep_tex: true
        includes:
            in_header: TeX_packages_commands.sty
---

<!-- ================================================================================================ -->
<!--   Beginning of Preamble : Preamble seldom requires change                                        -->
<!-- ================================================================================================ -->

<!-- knitr setup -->
```{r knitr setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=8,
  fig.height=8,
  fig.align = "center"
  )
```

<!-- R packages -->
```{r R packages, message=FALSE}
### ================
###   R packages
### ================
library(plyr)
library(dplyr)
library(qiimer)
library(pander)
library(ape)
library(vegan)
library(ggplot2)
library(tidyr)
library(broom)
library(reshape2)
library(kylemisc)
library(gtools) 
```

<!-- resources -->
```{r resources}
### ================
###   R resources
### ================
source("R_functions.R")
```

<!-- user defined functions -->
```{r user defined functions}
### ================
###   User defined functions, if any
### ================
```

<!-- ================================================================================================ -->
<!--   End of Preamble                                                                                -->
<!-- ================================================================================================ -->

<!-- ================================================================================================ -->
<!--   Beginning of Project Specific Items                                                            -->
<!-- ================================================================================================ -->

```{r}
### ===========================
###   define constants
### ===========================

### run date
run_date <- "PLEASE FILL ME IN WITH A DATE SUCH AS October 3rd, 2017"

### investigator
investigator <- "PLEASE FILL ME IN WITH INVESTIGATOR NAME SUCH AS Mark Brown"

### minimum reads threshold
min_reads <- 1000

### rarefying subsample size 
richness_subsample_size <- 1000

### data files directory
data_files_dir <- "data_files"

### porject directory
proj_dir <- "16s"

### mapping file path
#mapping_file_fp <- file.path(data_files_dir, "PLEASE FILL ME IN WITH MAPPING FILE NAME SUCH AS 20171003_brown_mapping_file.txt")
mapping_file_fp <- file.path(data_files_dir, "20171003_brown_mapping_file.txt")

### otu table file path
feature_table_fp <- file.path(proj_dir, "table", "feature-table.tsv")

### taxonomic assignment 
taxo_assignment_fp <- file.path(proj_dir, "taxonomy", "taxonomy.tsv")

### unweighted UniFrac file path
uu_fp <- file.path(proj_dir, "core-metrics-results", "uu", "distance-matrix.tsv")

### weighted UniFrac file path
wu_fp <- file.path(proj_dir, "core-metrics-results", "wu", "distance-matrix.tsv")

### faith
faith_fp <- file.path(proj_dir, "core-metrics-results", "faith", "alpha-diversity.tsv")
```

```{r, warning=F}
### ===========================
###   read in data
### ===========================

### read mapping file
s <- read_qiime_mapping_file(mapping_file_fp) 

### check for the column names to assign color_by and shape_by for pcoa plots
color_by <- NULL
shape_by <- NULL
potential_headers <- c("study_group", "SampleType", "study_day", "cage_number")
header_idx = which(is.element(potential_headers, colnames(s)))
if(length(header_idx)>0) {
  color_by <- potential_headers[header_idx[1]]
}
if(length(header_idx)>1) {
  shape_by <- potential_headers[header_idx[2]]
}

### Check if the sample names start with a number. If they do, add an "X" to the front
toFix = grepl("^[0-9]", s$SampleID)
s$SampleIDorig = s$SampleID
s$SampleID[toFix] = paste0('X', s$SampleID[toFix])
s$SampleID <- gsub("-", ".", s$SampleID)

### read otu table
o <- read.table(feature_table_fp, skip=1, header=T, as.is=T, sep="\t", comment.char="")
rownames(o) <- o$X.OTU.ID
counts <- o %>% select(-X.OTU.ID) %>% as.matrix()

### get read counts
read_counts <- colSums(counts) %>% 
  melt() %>% 
  dplyr::rename(Read_Counts=value) 
s <- merge(s, read_counts, by.x="SampleID", by.y="row.names", all.x=T)

### taxonomic assignment
ta <- read.table(file=taxo_assignment_fp, header=T, sep="\t", as.is=T)
ta$trunc_taxon <- sub("(; [kpcofgs]__)+$", "", ta$Taxon, perl=T)
rownames(ta) <- ta$Feature.ID
ta <- ta[rownames(o), c("trunc_taxon", "Confidence")]
md <- ta$trunc_taxon
names(md) <- rownames(ta)
adf <- split_assignments(md) 

### remove contamination
is_mitochondrial <- grepl("mitochondria", adf$Family)
is_chloroplast <- grepl("Chloroplast", adf$Class)
is_unassigned <- grepl("Unassigned", adf$Kingdom)
is_archaea <- grepl("Archaea", adf$Kingdom)
is_contam <- is_mitochondrial | is_chloroplast | is_unassigned ### Archaea kept to check positive control samples
counts <- counts[!is_contam,]
adf <- adf[!is_contam,]
rm(is_contam, is_mitochondrial, is_chloroplast, is_unassigned, is_archaea)

### alpha diversity measures: richness and shannon
s <- s %>%
  merge(diversity(t(counts)), by.x="SampleID", by.y="row.names", all.x=T) %>%
  dplyr::rename(shannon = y) %>%
  merge(rarefy(t(counts), richness_subsample_size), by.x="SampleID", by.y="row.names", all.x=T) %>%
  dplyr::rename(richness = y) 

### alpha diversity measures: faith
faith <- read.table(file=faith_fp, header=T, sep="\t", as.is=T) %>%
  dplyr::rename(SampleIDorig=X)
s <- merge(s, faith, by="SampleIDorig", all.x=T)
```

```{r}
### ===========================
###   data preprocessing and etc
### ===========================

### possible issue 1: any mismatch between s$SampleID and colames(counts)?
#in_s_not_in_counts <- setdiff(s$SampleID, colnames(counts))
#if (length(in_s_not_in_counts) > 0) stop (simpleError("These SampleID(s) are in the mapping file, but not found in the feature table.", paste(in_s_not_in_counts, collapse=" ")))
#in_counts_not_in_s <- setdiff(colnames(counts), s$SampleID)
#if (length(in_counts_not_in_s) > 0) stop (simpleError("These SampleID(s) are in the feature table, but not found in the mapping file.", paste(in_counts_not_in_s, collapse=" ")))
common_sample_list <- intersect(s$SampleID, colnames(counts))
s <- s[s$SampleID %in% common_sample_list, ]
counts <- counts[, common_sample_list]
```

# Introduction
This report is based on the results of sequencing performed on `r run_date` for `r investigator` Project. 

## Histogram of high quality paired reads per sample
The black dashed vertical line shows the minimum number of reads (`r min_reads`) for analysis. Control samples, if any, were included in the histogram.

```{r, fig.width=6, fig.height=4}
ggplot(s, aes(x=Read_Counts)) +
    geom_histogram(aes(fill=SampleType), binwidth=1000) +
    geom_vline(xintercept = min_reads, color="black", linetype="dashed") +
    theme_classic() +
    theme_bw() + 
    xlab("Number of reads in sample") +
    ylab("Number of samples")
```

\newpage

## Whole samples that are above the `r min_reads` read count threshold

```{r}
### Whole samples discarded that are below the read count threshold of min_reads
s$Keep <- s$Read_Counts > min_reads
s$KeepLabel <- factor(ifelse(s$Keep, "Keep", "Discard"))
keep_table <- table(s[, color_by], s$KeepLabel)
pander(keep_table)
```

\newpage

# Taxonomic heatmap

```{r}
prop_cut <- 0.01
satu_limit <- 0.4
```

Each column of the heatmap represents one sample and each row represents one taxon, typically a genus. Taxa were included in the chart if the abundance in any sample exceeded `r 100*prop_cut`%. 

The chart is colored white if taxa were not observed in the sample, dark blue if taxa were observed at very low abundance. This allows the reader to quickly survey presence/absence. Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

```{r, fig.height=12, fig.width=16}
s_toPlot <- s[s$Keep,] %>%
  mutate(study_group=as.character(study_group))
cts_toPlot <- counts[, s_toPlot$SampleID]  
a <- simplify_assignments(adf, rank1="Phylum", rank2="Genus")
summed_cts <- rowsum(cts_toPlot, a) 
summed_props <- sweep(summed_cts, 2, colSums(summed_cts), "/") 
grps <- c(color_by, shape_by)

### heatmap with a gap
heatmap_grouped(summed_props, s_toPlot, grps=grps, thre=0.01, option=2, prop_cut = prop_cut, satu_limit=satu_limit)
```

\newpage

# Alpha Diversity

Alpha diversity was assessd by the expected number of observed OTUs (out of rarefying sample size of `r richness_subsample_size`), Shannon index, and Faith’s phylogenetic diversity.

## Number of observed OTUs

```{r, fig.height=2.5}
alpha_measure <- "richness"

s %>% filter(Keep) %>%
  ggplot(aes(x=eval(parse(text=color_by)), y=eval(parse(text=alpha_measure)), color=eval(parse(text=color_by)))) +
  geom_boxplot() +
  labs(y=alpha_measure, x=color_by, color=color_by) +
  theme_bw() +
  theme(axis.text.x=element_text(angle=-25, hjust= .1)) +
  scale_color_brewer(palette = "Set2")
```

## Shannon Index

```{r, fig.height=2.5}
alpha_measure <- "shannon"

s %>% filter(Keep) %>%
  ggplot(aes(x=eval(parse(text=color_by)), y=eval(parse(text=alpha_measure)), color=eval(parse(text=color_by)))) +
  geom_boxplot() +
  labs(y=alpha_measure, x=color_by, color=color_by) +
  theme_bw() +
  theme(axis.text.x=element_text(angle=-25, hjust= .1)) +
  scale_color_brewer(palette = "Set2")
```

## Faith’s Phylogenetic Diversity

```{r, fig.height=2.5}
alpha_measure <- "faith_pd"

s %>% filter(Keep) %>%
  ggplot(aes(x=eval(parse(text=color_by)), y=eval(parse(text=alpha_measure)), color=eval(parse(text=color_by)))) +
  geom_boxplot() +
  labs(y=alpha_measure, x=color_by, color=color_by) +
  theme_bw() +
  theme(axis.text.x=element_text(angle=-25, hjust= .1)) +
  scale_color_brewer(palette = "Set2")
```

\newpage 

# Beta diversity

Similarity between samples was assessed by unweighted and weighted UniFrac distances. 

## Unweighted UniFrac distances

```{r}
dist_type <- "uu" ## unweighted or weighted
if (dist_type=="uu") {
  dist_fp <- uu_fp
  dist_name <- "unweighted UniFrac distances"
} else if (dist_type=="wu") {
  dist_fp <- wu_fp
  dist_name <- "weighted UniFrac distances"
}
```

The `r dist_name` was employed to compare the species composition of the samples to each other. We used a method of ordination called Principal Coordinates Analysis to select the best 2D coordinate system for display. The percentage in each axis represents the proportion of the total variance captured along the axis. 

### PCoA plot

```{r, fig.width=6, fig.height=5, echo=FALSE, warning=FALSE}
s_toPlot <- s[s$Keep, ]
s_toPlot$Keep <- ifelse(grepl("geneblock|freewater|extractemptywell", s_toPlot$SampleID, ignore.case = TRUE), FALSE, TRUE)

dist_in <- read_qiime_distmat(dist_fp)
dist_in <- dist_subset(dist_in, s_toPlot$SampleID[s_toPlot$Keep])
pcoa_title <- paste0("PCoA plot based on ", dist_name)
pcoa_plot <- make_pcoa_plot(dist_in, s_toPlot[s_toPlot$Keep,], color_by=color_by, shape_by=NULL, title = pcoa_title)
plot(pcoa_plot)
```

\newpage

### UPGMA clustering 

The following plot shows sample clustering based on `r dist_name`. We have used a method of hierarchical clustering called "average-linkage" or UPGMA. At the bottom of the dendrogram, all samples start out in their own group. Moving up the dendrogram, samples accumulate into clusters if the average (mean) distance between all samples is below the indicated value.

```{r, fig.width=12, fig.height=5, echo=FALSE}
hc = hclust(dist_in, method="average")
plot(hc, main=paste0("UPGMA linkage clustergram based on\n", dist_name), xlab="", ylab="distance", sub = "")
```

\newpage

## Weighted UniFrac distances

```{r}
dist_type <- "wu" ## unweighted or weighted
if (dist_type=="uu") {
  dist_fp <- uu_fp
  dist_name <- "unweighted UniFrac distances"
} else if (dist_type=="wu") {
  dist_fp <- wu_fp
  dist_name <- "weighted UniFrac distances"
}
```

The `r dist_name` was employed to compare the species composition of the samples to each other. We used a method of ordination called Principal Coordinates Analysis to select the best 2D coordinate system for display. The percentage in each axis represents the proportion of the total variance captured along the axis. 

### PCoA plot

```{r, fig.width=6, fig.height=5, echo=FALSE, warning=FALSE}
s_toPlot <- s[s$Keep, ]
s_toPlot$Keep <- ifelse(grepl("geneblock|freewater|extractemptywell", s_toPlot$SampleID, ignore.case = TRUE), FALSE, TRUE)

dist_in <- read_qiime_distmat(dist_fp)
dist_in <- dist_subset(dist_in, s_toPlot$SampleID[s_toPlot$Keep])
pcoa_title <- paste0("PCoA plot based on ", dist_name)
pcoa_plot <- make_pcoa_plot(dist_in, s_toPlot[s_toPlot$Keep,], color_by=color_by, shape_by=NULL, title = pcoa_title)
plot(pcoa_plot)
```

\newpage

### UPGMA clustering 

The following plot shows sample clustering based on `r dist_name`. We have used a method of hierarchical clustering called "average-linkage" or UPGMA. At the bottom of the dendrogram, all samples start out in their own group. Moving up the dendrogram, samples accumulate into clusters if the average (mean) distance between all samples is below the indicated value.

```{r, fig.width=12, fig.height=5, echo=FALSE}
hc = hclust(dist_in, method="average")
plot(hc, main=paste0("UPGMA linkage clustergram based on\n", dist_name), xlab="", ylab="distance", sub = "")
```

\newpage

# APPENDIX: Counts of high quality paired reads for each sample
```{r}
s %>% select(SampleID, Read_Counts) %>% arrange(-Read_Counts) %>% pander()
```
